### PR TITLE
Implement terms and conditions agreement

### PR DESCRIPTION
**Pull request by richardpattinson**
### Context

Introduce a system prompt for users to agree to the service's terms and conditions upon their first sign-in and at periodic intervals thereafter to ensure ongoing compliance and user understanding of the service policies. The T&C document is ready for implementation.

Objectives:

Ensure Compliance: Guarantee that all users are aware of and agree to the terms and conditions governing the use of the service.

User Awareness: Enhance user understanding of service policies, including data use, privacy, and user responsibilities.

Periodic Acknowledgment: Require users to periodically review and agree to the terms and conditions to account for updates or changes.

### Checklist

- [x] Attach to Trello card
- [x] Rebased main
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: app/controllers/concerns/dsi_authenticatable.rb [hunk 1]
```diff
@@ -4,8 +4,11 @@ module DsiAuthenticatable
   included do
     before_action :authenticate_dsi_user!
     before_action :handle_expired_session!
+    before_action :enforce_terms_and_conditions_acceptance!
```
Comment (malcolmbaig) [inline app/controllers/concerns/dsi_authenticatable.rb]: I don't think we should conflate authentication concerns with the terms and conditions flow. We could alternatively put T&Cs logic in the CheckRecordsController where it'll still apply to all CTR requests, and make the check only if `current_dsi_user` exists.

File: app/controllers/concerns/dsi_authenticatable.rb [hunk 2]
```diff
@@ -16,6 +19,12 @@ def authenticate_dsi_user!
     end
   end
 
+  def enforce_terms_and_conditions_acceptance!
+    if FeatureFlags::FeatureFlag.active?("terms_and_conditions") && current_dsi_user.terms_and_conditions_version_accepted != CURRENT_TERMS_AND_CONDITIONS_VERSION
+        redirect_to check_records_terms_and_conditions_path
+    end
```
Comment (malcolmbaig) [inline app/controllers/concerns/dsi_authenticatable.rb]: This check needs updating so we re-prompt them every six months (or whatever value was agreed).

File: app/controllers/check_records/terms_and_conditions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+module CheckRecords
+  class TermsAndConditionsController < CheckRecordsController
+    def show
+    end
+
+    def update
+      current_dsi_user.update!(terms_and_conditions_version_accepted: CURRENT_TERMS_AND_CONDITIONS_VERSION,
+                              terms_and_conditions_timestamp: Time.zone.now)
+      redirect_to check_records_root_path notice: "Terms and conditions accepted"
```
Comment (malcolmbaig) [inline app/controllers/check_records/terms_and_conditions_controller.rb]:
```suggestion
      redirect_to check_records_root_path, notice: "Terms and conditions accepted"
```

File: config/routes/check_records.rb [hunk 4]
```diff
@@ -13,6 +13,9 @@
   get "/auth/dfe/callback", to: "omniauth_callbacks#dfe"
   post "/auth/developer/callback" => "omniauth_callbacks#dfe_bypass"
 
+  get "/terms-and-conditions", to: "terms_and_conditions#show"
+  patch "/auth/developer/callback" => "terms_and_conditions#update"
```
Comment (malcolmbaig) [inline config/routes/check_records.rb]: patch path looks incorrect here.

File: spec/system/check_records/user_accepts_terms_and_conditions_spec.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  include ActivateFeaturesSteps
+
+  scenario "User accepts terms and conditions", test: :with_stubbed_auth do
+    given_the_check_service_is_open
+    given_the_terms_and_conditions_feature_flag_is_enabled
+    when_i_sign_in_via_dsi
+    then_i_am_signed_in
+    and_i_am_redirected_to_the_terms_and_conditions_page
+    then_i_click_accept
+    then_i_am_taken_to_the_root
```
Comment (malcolmbaig) [inline spec/system/check_records/user_accepts_terms_and_conditions_spec.rb:25]: Let's also assert that we see the flash notice (will validate the suggested change [here](https://github.com/DFE-Digital/access-your-teaching-qualifications/pull/679/files#r1647343820)), and confirm the database changes on the dsi_user record.

**Review by malcolmbaig — CHANGES_REQUESTED** [review]
A few comments left inline. After they're addressed, would you mind testing this locally and adding screenshots to the PR description?

File: app/controllers/check_records/check_records_controller.rb [hunk 6]
```diff
@@ -16,5 +19,13 @@ class CheckRecordsController < ApplicationController
     def current_namespace
       "check-a-teachers-record"
     end
+
+    def enforce_terms_and_conditions_acceptance!
+      if current_dsi_user &&
+        (FeatureFlags::FeatureFlag.active?("terms_and_conditions") &&
+          current_dsi_user.terms_and_conditions_version_accepted != CURRENT_TERMS_AND_CONDITIONS_VERSION)
+          redirect_to check_records_terms_and_conditions_path
```
Comment (malcolmbaig) [inline app/controllers/check_records/check_records_controller.rb]: There's still no logical check here based on the datetime they accepted the agreement.

File: config/routes/check_records.rb [hunk 7]
```diff
@@ -18,6 +18,12 @@
   get "/trn-search", to: "search#trn_search"
   get "/trn-result", to: "search#trn_result"
 
+  get "/terms-and-conditions", to: "terms_and_conditions#show"
+  patch "/terms-and-conditions" => "terms_and_conditions#update"
+
+  get "/search", to: "search#new"
+  get "/result", to: "search#show"
```
Comment (malcolmbaig) [inline config/routes/check_records.rb]: Looks like a merge conflict from the search results work.

Comment (malcolmbaig) [inline config/routes/check_records.rb]:
```suggestion
```

File: spec/system/check_records/user_accepts_terms_and_conditions_spec.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    and_i_see_the_successful_notification
+    and_the_dsi_user_has_been_updated
```
Comment (malcolmbaig) [inline spec/system/check_records/user_accepts_terms_and_conditions_spec.rb:27]: Once the timestamp check has also been added, this spec can be expanded to travel forward in time and assert that the user is prompted again.

File: app/controllers/check_records/check_records_controller.rb [hunk 9]
```diff
@@ -16,5 +19,24 @@ class CheckRecordsController < ApplicationController
...
+    def acceptance_required?
+      !current_version_accepted || acceptance_expired
+    end
+
+    def current_version_accepted
+      current_dsi_user.terms_and_conditions_version_accepted == CURRENT_TERMS_AND_CONDITIONS_VERSION
+    end
+
+    def acceptance_expired
+      current_dsi_user.terms_and_conditions_timestamp > 12.months.ago
```
Comment (malcolmbaig) [inline app/controllers/check_records/check_records_controller.rb]: To confirm - it's 12 months and not 6?

File: spec/system/check_records/user_accepts_terms_and_conditions_spec.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    when_13_months_has_passed
+    when_i_go_to_the_root
+    and_i_am_redirected_to_the_terms_and_conditions_page
+    then_i_click_accept
+    then_i_am_taken_to_the_root
+    and_i_see_the_successful_notification
+    and_the_dsi_user_has_been_updated
```
Comment (malcolmbaig) [inline spec/system/check_records/user_accepts_terms_and_conditions_spec.rb]: nit: can remove the last step, not testing anything with its addition.

File: db/schema.rb [hunk 11]
```diff
@@ -117,6 +117,8 @@
     t.string "uid", null: false
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.string "terms_and_conditions_version_accepted"
+    t.datetime "terms_and_conditions_timestamp"
```
Comment (felixclack) [inline db/schema.rb]: Rails conventions are to use an `_at` suffix for timestamp fields. How about applying that here?

eg.

```suggestion
    t.datetime "terms_and_conditions_accepted_at"
```

File: Gemfile.lock [hunk 12]
```diff
@@ -276,7 +276,7 @@ GEM
       bindata
       faraday (~> 2.0)
       faraday-follow_redirects
-    jwt (2.8.1)
+    jwt (2.8.2)
```
Comment (felixclack) [inline Gemfile.lock]: Looks like some unrelated changes have snuck in here.

File: app/controllers/check_records/check_records_controller.rb [hunk 13]
```diff
@@ -16,5 +18,27 @@ class CheckRecordsController < ApplicationController
...
+      !current_version_accepted || acceptance_expired
+    end
+
+    def current_version_accepted
+      current_dsi_user.terms_and_conditions_version_accepted == CURRENT_TERMS_AND_CONDITIONS_VERSION
+    end
+
+    def acceptance_expired
+      current_dsi_user.terms_and_conditions_timestamp < 12.months.ago
+    end
```
Comment (felixclack) [inline app/controllers/check_records/check_records_controller.rb]: This is business logic and would be a good fit for being implemented within a model.

Either as a method on user or as an object that accepts the relevant fields and returns a boolean.

eg.

```
# app/models/user.rb
class User
  ...
  
  CURRENT_TERMS_AND_CONDITIONS_VERSION = "1.0".freeze

  def acceptance_required?
    !current_version_accepted? || acceptance_expired?
  end
  
  private

  def current_version_accepted?
    terms_and_conditions_version_accepted == CURRENT_TERMS_AND_CONDITIONS_VERSION
  end

  def acceptance_expired?
    terms_and_conditions_at.before?(12.months.ago)
  end
end
```

Or as a separate model
```
# app/models/terms_expiry.rb
class TermsExpiry
  include ActiveModel::Model
  CURRENT_TERMS_AND_CONDITIONS_VERSION = "1.0".freeze

  attr_accessor :accepted_at, :version

  validates :accepted_at, presence: true
  validates :version, presence: true

  def acceptance_required?
    !current_version_accepted? || acceptance_expired
  end

  private

  def current_version_accepted?
    version == CURRENT_TERMS_AND_CONDITIONS_VERSION
  end

  def acceptance_expired?
    accepted_at.before?(12.months.ago)
  end
end
```

File: app/controllers/check_records/terms_and_conditions_controller.rb [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+module CheckRecords
+  class TermsAndConditionsController < CheckRecordsController
+    def show
+    end
+
+    def update
+      current_dsi_user.update!(
+        terms_and_conditions_version_accepted: CURRENT_TERMS_AND_CONDITIONS_VERSION,
+        terms_and_conditions_timestamp: Time.zone.now
+      )
```
Comment (felixclack) [inline app/controllers/check_records/terms_and_conditions_controller.rb]: I would define a method for accepting terms and keep this logic out of the controller.

eg. `current_dsi_user.accept_terms!`

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 15]
```diff
@@ -0,0 +1,210 @@
+<% content_for :page_title, "Terms and Conditions" %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <h1 class="govuk-heading-l">Terms and Conditions    </h1>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]:
```suggestion
    <h1 class="govuk-heading-l">Terms and Conditions</h1>
```

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+
+    <p class="govuk-body">1.3 Please read these Terms carefully before You access and start to use the Service as they tell You the rules for using the Service.</p>
+
+    <p class="govuk-body">1.4 By accessing and using the Service, You confirm that You accept and are bound by the Terms and have a valid basis for accessing and using the Service. If You do not agree to these Terms, You are not permitted to use this Service and You should not attempt to do so.</p>
+
+    <p class="govuk-body">1.5 These Terms refer to the following additional terms, which also apply to your use of the Service: </p>
+    <ol type="a" class="govuk-body">
+        <li class="govuk-body">our privacy notice that details the processing of Personal Data through the Service (as may be amended by us from time to time);</li>
+
+        <li class="govuk-body">any <a class="govuk-link" href="https://interactions.signin.education.gov.uk/terms">DfE Sign-In</a> terms and conditions that govern the access to the Service (as such terms of use may be amended by us from time to time).</li>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]:
```suggestion
      <li class="govuk-body">our privacy notice that details the processing of Personal Data through the Service (as may be amended by us from time to time);</li>

      <li class="govuk-body">any <a class="govuk-link" href="https://interactions.signin.education.gov.uk/terms">DfE Sign-In</a> terms and conditions that govern the access to the Service (as such terms of use may be amended by us from time to time).</li>
```

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+    <h2 class ="govuk-heading-m">
+      3. Terms of Use
+    </h2>
+    <p class="govuk-body">3.1 These Terms must be complied with whenever You access the Service. You acknowledge that in addition to complying with these Terms, You must also comply with all applicable law, including Data Protection Legislation.</p>
+
+    <p class="govuk-body">3.2 In accessing the Service, You also acknowledge your organisation is an independent controller, as defined in Data Protection Legislation, with responsibility for ensuring:</p>
+    <ol type="a" class="govuk-body">
+        <li class="govuk-body">your use of data obtained from the Service complies with Data Protection Legislation; and</li>
+
+        <li class="govuk-body">any processors required to access the Service on your organisation’s behalf do so with a valid basis for processing and always use any Personal Data obtained from the Service in compliance with Data Protection Legislation and in accordance with these Terms.</li>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]:
```suggestion
      <li class="govuk-body">your use of data obtained from the Service complies with Data Protection Legislation; and</li>

      <li class="govuk-body">any processors required to access the Service on your organisation’s behalf do so with a valid basis for processing and always use any Personal Data obtained from the Service in compliance with Data Protection Legislation and in accordance with these Terms.</li>
```

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+    </ol>
+    <p class="govuk-body">3.3 Without prejudice to Clause 3.2 and as set out in Clause 1.6, You will also need to comply with the Annex to these Terms in circumstances where your organisation does not have a separate data sharing agreement or memorandum of understanding (as applicable) in place with DfE.</p>
+
+    <p class="govuk-body">3.4 As a user of this Service, You will at all times:</p>
+    <ol type="a" class="govuk-body">
+
+        <li class="govuk-body">comply with the provisions of Data Protection Legislation in respect of all Personal Data, understanding that such Personal Data is provided to enable only authorised users from a registered organisation to carry out necessary checks on whether a trainee or teacher has:</li>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]:
```suggestion
      <li class="govuk-body">comply with the provisions of Data Protection Legislation in respect of all Personal Data, understanding that such Personal Data is provided to enable only authorised users from a registered organisation to carry out necessary checks on whether a trainee or teacher has:</li>
```

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+                <li class="govuk-body">achieved qualified teacher status, early years teacher status or early years professional status;</li>
+
+                <li class="govuk-body">completed their induction;</li>
+
+                <li class="govuk-body">been awarded a mandatory qualification for teachers of hearing impaired or visually impaired pupils;</li>
+
+                <li class="govuk-body">successfully completed a leadership or specialist national professional qualification;</li>
+
+                <li class="govuk-body">any teaching restrictions placed against them or has been the subject of a decision by the Secretary of State for Education not to impose a prohibition order following a determination by a professional conduct panel of unacceptable professional conduct or conduct that may bring the teaching profession into disrepute or the conviction of a relevant offence.</li>
+            </ol>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]:
```suggestion
    <ol type="i" class="govuk-body">
      <li class="govuk-body">achieved qualified teacher status, early years teacher status or early years professional status;</li>

      <li class="govuk-body">completed their induction;</li>

      <li class="govuk-body">been awarded a mandatory qualification for teachers of hearing impaired or visually impaired pupils;</li>

      <li class="govuk-body">successfully completed a leadership or specialist national professional qualification;</li>

      <li class="govuk-body">any teaching restrictions placed against them or has been the subject of a decision by the Secretary of State for Education not to impose a prohibition order following a determination by a professional conduct panel of unacceptable professional conduct or conduct that may bring the teaching profession into disrepute or the conviction of a relevant offence.</li>
```

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+    <ol type="a">
+        <li class="govuk-body">takes no responsibility for the accuracy or completeness or the security of any personal data transferred to, used in or stored on your system or in your local files. In such circumstances, for the purposes of Data Protection Legislation, You understand that your organisation is the controller for any such personal data and any activities carried out with this data must comply with Data Protection Legislation.</li>
+
+        <li class="govuk-body">takes reasonable steps to ensure the quality, accuracy, and completeness of its data but accept that these factors cannot be guaranteed.</li>
+
+        <li class="govuk-body">reserves the right to reset user passwords.</li>
+
+        <li class="govuk-body">owns all intellectual property rights (which is protected by Crown copyright) in and to the Service which includes any and all data (including Personal Data) or other information or material made available or published via the Service and these rights are licensed (not sold) to You. You have no intellectual property rights in relation to the foregoing other than the right to use them in accordance with these Terms.</li>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]:
```suggestion
      <li class="govuk-body">takes no responsibility for the accuracy or completeness or the security of any personal data transferred to, used in or stored on your system or in your local files. In such circumstances, for the purposes of Data Protection Legislation, You understand that your organisation is the controller for any such personal data and any activities carried out with this data must comply with Data Protection Legislation.</li>

      <li class="govuk-body">takes reasonable steps to ensure the quality, accuracy, and completeness of its data but accept that these factors cannot be guaranteed.</li>

      <li class="govuk-body">reserves the right to reset user passwords.</li>

      <li class="govuk-body">owns all intellectual property rights (which is protected by Crown copyright) in and to the Service which includes any and all data (including Personal Data) or other information or material made available or published via the Service and these rights are licensed (not sold) to You. You have no intellectual property rights in relation to the foregoing other than the right to use them in accordance with these Terms.</li>
```

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+            </ol>
+
+    <li class="govuk-body">not disclose any Personal Data from the Service to any unauthorised third party.</li>
+
+    <li class="govuk-body">abide by all instructions, conditions and other requirements made known to You by us when accessing and using the Service.</li>
+
+    <li class="govuk-body">seek prior written advice from DfE where You are unsure whether a proposed use of the Service is permitted under Data Protection Legislation.</li>
+
+    <li class="govuk-body">indemnify and hold harmless DfE (and for the avoidance of doubt, the TRA) against any and all losses, liabilities, demands, fines, penalties, claims for compensation, costs (including without limitation legal costs), expenses or damages arising from any unauthorised use, loss or corruption of Personal Data available via the Service where the foregoing arises out of or in connection with any act or omission You have taken.</li>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]:
```suggestion
      <li class="govuk-body">not disclose any Personal Data from the Service to any unauthorised third party.</li>

      <li class="govuk-body">abide by all instructions, conditions and other requirements made known to You by us when accessing and using the Service.</li>

      <li class="govuk-body">seek prior written advice from DfE where You are unsure whether a proposed use of the Service is permitted under Data Protection Legislation.</li>

      <li class="govuk-body">indemnify and hold harmless DfE (and for the avoidance of doubt, the TRA) against any and all losses, liabilities, demands, fines, penalties, claims for compensation, costs (including without limitation legal costs), expenses or damages arising from any unauthorised use, loss or corruption of Personal Data available via the Service where the foregoing arises out of or in connection with any act or omission You have taken.</li>
```

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+    <ol type="a">
+        <li class="govuk-body">any "robot", "bot", "spider", "scraper" or other automated device, program, tool, algorithm, code, process or methodology to access, obtain, copy, monitor or republish any portion of the Service or any content or services accessed via the same; or</li>
+
+        <li class="govuk-body">any automated analytical technique aimed at analysing content in digital form to generate information which includes but is not limited to patterns, trends and correlations,</li>
+
+        <li class="govuk-body">the provisions in this Clause should be treated as an express reservation of our rights in this regard, including for the purposes of Article 4(3) of Digital Copyright Directive ((EU) 2019/790). This Clause shall not apply insofar as (but only to the extent that) we are unable to exclude or limit text or data mining or web scraping activity by contract under the laws which are applicable to us.</li>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]:
```suggestion
      <li class="govuk-body">any "robot", "bot", "spider", "scraper" or other automated device, program, tool, algorithm, code, process or methodology to access, obtain, copy, monitor or republish any portion of the Service or any content or services accessed via the same; or</li>

      <li class="govuk-body">any automated analytical technique aimed at analysing content in digital form to generate information which includes but is not limited to patterns, trends and correlations,</li>

      <li class="govuk-body">the provisions in this Clause should be treated as an express reservation of our rights in this regard, including for the purposes of Article 4(3) of Digital Copyright Directive ((EU) 2019/790). This Clause shall not apply insofar as (but only to the extent that) we are unable to exclude or limit text or data mining or web scraping activity by contract under the laws which are applicable to us.</li>
```

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+      9. Contact
+    </h2>
+
+    <p class="govuk-body">9.1 By registering to use the Service, You are giving DfE (via the TRA) permission to contact You or your organisation from time to time in relation to your access to and usage of the Service. We may review how you use the Service to ensure that any use by You is in accordance with these Terms.</p>
+
+    <p class="govuk-body">Please contact us if You have any queries regarding these Terms or your use of the Service via our contact details:</p>
+
+    <p class="govuk-body"><a class="govuk-link" mailto="teaching.status@education.gov.uk">Email:	teaching.status@education.gov.uk</a></p>
+
+    <p class="govuk-body">Website:	<a class="govuk-link" href="https://www.gov.uk/government/organisations/teaching-regulation-agency">https://www.gov.uk/government/organisations/teaching-regulation-agency</a></p>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]:
```suggestion
    <p class="govuk-body">Website: <a class="govuk-link" href="https://www.gov.uk/government/organisations/teaching-regulation-agency">https://www.gov.uk/government/organisations/teaching-regulation-agency</a></p>
```

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+
+    <h2 class ="govuk-heading-m">
+      9. Contact
+    </h2>
+
+    <p class="govuk-body">9.1 By registering to use the Service, You are giving DfE (via the TRA) permission to contact You or your organisation from time to time in relation to your access to and usage of the Service. We may review how you use the Service to ensure that any use by You is in accordance with these Terms.</p>
+
+    <p class="govuk-body">Please contact us if You have any queries regarding these Terms or your use of the Service via our contact details:</p>
+
+    <p class="govuk-body"><a class="govuk-link" mailto="teaching.status@education.gov.uk">Email:	teaching.status@education.gov.uk</a></p>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]:
```suggestion
    <p class="govuk-body"><a class="govuk-link" mailto="teaching.status@education.gov.uk">Email: teaching.status@education.gov.uk</a></p>
```

**Comment by felixclack** [comment]
What alternative solutions for storing acceptance of terms did you consider here?

File: app/controllers/check_records/check_records_controller.rb (hunk 13 above)
Comment (richardpattinson) [reply app/controllers/check_records/check_records_controller.rb]: I've put the logic in the dsi_users model

**Comment by richardpattinson** [comment]
> What alternative solutions for storing acceptance of terms did you consider here?

I considered adding a new terms and conditions acceptance table that had foreign keys to a version number and a user. I went with this method as requires less storage, though both were pretty minimal. What would you reccomend for storing this type of information?

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+          <li class="govuk-body">achieved qualified teacher status, early years teacher status or early years professional status;</li>
+
+          <li class="govuk-body">completed their induction;</li>
+
+          <li class="govuk-body">been awarded a mandatory qualification for teachers of hearing impaired or visually impaired pupils;</li>
+
+          <li class="govuk-body">successfully completed a leadership or specialist national professional qualification;</li>
+
+          <li class="govuk-body">any teaching restrictions placed against them or has been the subject of a decision by the Secretary of State for Education not to impose a prohibition order following a determination by a professional conduct panel of unacceptable professional conduct or conduct that may bring the teaching profession into disrepute or the conviction of a relevant offence.</li>
+        </ol>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]: Is this supposed to be a separate list? It is invalid HTML in its current form.

File: app/views/check_records/terms_and_conditions/show.html.erb [hunk 26]
```diff
@@ -0,0 +1,210 @@
...
+
+      <li class="govuk-body">indemnify and hold harmless DfE (and for the avoidance of doubt, the TRA) against any and all losses, liabilities, demands, fines, penalties, claims for compensation, costs (including without limitation legal costs), expenses or damages arising from any unauthorised use, loss or corruption of Personal Data available via the Service where the foregoing arises out of or in connection with any act or omission You have taken.</li>
+    </ol>
+    <p class="govuk-body">3.5 As a user of this Service, You understand DfE:</p>
+    <ol type="a">
+        <li class="govuk-body">takes no responsibility for the accuracy or completeness or the security of any personal data transferred to, used in or stored on your system or in your local files. In such circumstances, for the purposes of Data Protection Legislation, You understand that your organisation is the controller for any such personal data and any activities carried out with this data must comply with Data Protection Legislation.</li>
```
Comment (felixclack) [inline app/views/check_records/terms_and_conditions/show.html.erb]: Indentation looks inconsistent here compared to the rest of the file. It is using 4 spaces whereas we use 2 spaces everywhere else.

**Comment by felixclack** [comment]
> > What alternative solutions for storing acceptance of terms did you consider here?
> 
> I considered adding a new terms and conditions acceptance table that had foreign keys to a version number and a user. I went with this method as requires less storage, though both were pretty minimal. What would you reccomend for storing this type of information?

I don't know the full scope of what was asked for, so I'm making some assumptions here. Assumption one is that we want a log of all acceptance of terms over time. Secondly, we want an arbitrary time period for the expiration of acceptance. Lastly, I assume storage is cheap and only getting cheaper.

The other thing to consider here and anytime you create something related to the user. In most systems, the `User` model gets overloaded over time through adding small incremental features like this. This ends up turning `User` into a god-model that is tightly coupled to many different domains in the app.

A way to avoid this is by ensuring you only add things to the `User` related to identity. eg. some kind of unique identifier plus the auth stuff.

With those assumptions in mind I would consider the benefits of modelling acceptance as a separate entity.

eg.

```ruby
class TermsAcceptance
  VERSION = "1.0".freeze
  FUTURE = Time.current..
  EXPIRY_PERIOD = 12.months

  belongs_to :user

  scope :not_expired, -> { where(expires_at: FUTURE) }

  def accept
    update(
      expires_at: EXPIRY_PERIOD.from_now
      version: VERSION
    )
  end
end

class User
  has_many :terms_acceptances
  has_many :valid_terms_acceptances, -> { not_expired.order(created_at: :desc) }, class: 'TermsAcceptance'
end
```

Then when we create acceptance of terms in a controller like:

```ruby
class TermAcceptancesController
  def new
    @terms = current_dsi_user.terms_acceptances.build
  end

  def create
    @terms = current_dsi_user.terms_acceptances.build
    if terms.accept
      redirect_to check_records_root_path, notice: 'Terms accepted'
    else
      render :new
    end
  end
end

# app/views/terms_acceptances/new.html.erb
...
<%= form_with model: @terms do |form| %>
  <%= govuk_submit 'Accept' %>
<% end %>

# config/routes.rb
get '/terms-and-conditions', to: 'terms_acceptances#new'
post '/terms-and-conditions', to: 'terms_acceptances#create'
```

This is not a deal breaker for me as I think we have a pressing need to ship the T&Cs and I don't want to delay that.

Consider this an exercise in considering the alternatives and how that might improve the system design in the future.

We can always come back and refactor this in the future.

File: app/controllers/check_records/check_records_controller.rb [hunk 27]
```diff
@@ -16,5 +18,16 @@ class CheckRecordsController < ApplicationController
     def current_namespace
       "check-a-teachers-record"
     end
+
+    def enforce_terms_and_conditions_acceptance!
+      return if request.path == check_records_terms_and_conditions_path
+      return if request.path == check_records_sign_in_path
+      return if request.path == check_records_sign_out_path
```
Comment (felixclack) [inline app/controllers/check_records/check_records_controller.rb]: Putting this routing logic here means the method is violating the single responsibility principle. I would consider moving that logic to a skip_before_action or conditionally include this where relevant.

Also, this code is untested, let's add a test for it.

Comment (richardpattinson) [reply app/controllers/check_records/check_records_controller.rb]: Ive added some lines that check that we cannot navigate away from the terms and conditions page and have moved these guard conditions into skip_before_actions on the relevant controllers. Is a seperate system spec needed for checking that sign_in and out are still possible without approving the t&c's?

File: Gemfile.lock [hunk 28]
```diff
@@ -278,7 +278,7 @@ GEM
       bindata
       faraday (~> 2.0)
       faraday-follow_redirects
-    jwt (2.8.2)
+    jwt (2.8.1)
```
Comment (felixclack) [inline Gemfile.lock]: This change appears to be unrelated to terms and conditions.

File: app/controllers/check_records/check_records_controller.rb [hunk 29]
```diff
@@ -16,5 +18,12 @@ class CheckRecordsController < ApplicationController
     def current_namespace
       "check-a-teachers-record"
     end
+
+    def enforce_terms_and_conditions_acceptance!
+      if current_dsi_user &&
+        (FeatureFlags::FeatureFlag.active?("terms_and_conditions") && current_dsi_user.acceptance_required?)
```
Comment (felixclack) [inline app/controllers/check_records/check_records_controller.rb]: Do we need the feature flag?

Comment (felixclack) [inline app/controllers/check_records/check_records_controller.rb]:
```suggestion
      if current_dsi_user&.acceptance_required?
```

File: app/controllers/concerns/dsi_authenticatable.rb [hunk 30]
```diff
@@ -6,6 +6,8 @@ module DsiAuthenticatable
     before_action :handle_expired_session!
   end
 
+  CURRENT_TERMS_AND_CONDITIONS_VERSION = "1.0".freeze
```
Comment (felixclack) [inline app/controllers/concerns/dsi_authenticatable.rb]: This isn't needed anymore

File: app/views/check_records/terms_and_conditions/show.html.erb (hunk 25 above)
Comment (felixclack) [reply app/views/check_records/terms_and_conditions/show.html.erb]: This is invalid HTML and still needs addressing.

File: db/migrate/20240709072134_rename_terms_and_conditions_timestamp_to_terms_and_conditions_accepted_at.rb [hunk 31]
```diff
@@ -0,0 +1,6 @@
+class RenameTermsAndConditionsTimestampToTermsAndConditionsAcceptedAt < ActiveRecord::Migration[7.1]
+  def change
+    rename_column :dsi_users, :terms_and_conditions_timestamp, :terms_and_conditions_accepted_at
+
```
Comment (felixclack) [inline db/migrate/20240709072134_rename_terms_and_conditions_timestamp_to_terms_and_conditions_accepted_at.rb]: Unnecessary blank line.

```suggestion
```

**Review by felixclack — CHANGES_REQUESTED** [review]
Left some comments to fix a few things up.

File: config/feature_flags.yml [hunk 32]
```diff
@@ -36,3 +36,6 @@ feature_flags:
     author: Malcolm Baig
     description: Reduce number of search results shown to users of CTR by prompting
       them for a TRN when last name and date of birth return multiple results.
+  terms_and_conditions:
+    author: Richard Pattinson
+    description: Check whether a user has accepted the latest version of the terms and conditions.
```
Comment (felixclack) [inline config/feature_flags.yml]: Not needed anymore as you have removed the feature flag.

File: spec/support/system/activate_features_steps.rb [hunk 33]
```diff
@@ -7,6 +7,10 @@ def given_the_qualifications_service_is_open
     FeatureFlags::FeatureFlag.activate(:qualifications_service_open)
   end
 
+  def given_the_terms_and_conditions_feature_flag_is_enabled
+    FeatureFlags::FeatureFlag.activate(:terms_and_conditions)
+  end
+
```
Comment (felixclack) [inline spec/support/system/activate_features_steps.rb]: Not needed anymore

Comment (richardpattinson) [reply spec/support/system/activate_features_steps.rb]: Updated

File: config/feature_flags.yml (hunk 32 above)
Comment (richardpattinson) [reply config/feature_flags.yml]: Updated

**Review by felixclack — APPROVED** [review]
